File: bridges/policy.py
"""Trust policy for installing MCP servers and mounting code-running bridges.

The kernel needs to draw a line: "code from this source can run silently;
code from that source has to ask first." This is that line.

policy modes:
  - "auto":   trusted sources/kinds run silently; unknown sources prompt.
  - "prompt": every install / risky mount prompts (the conservative default
              before this policy existed).
  - "off":    installs and code-running bridge mounts are refused.

Risky kinds (those that exec arbitrary commands) — `mcp`, `process`,
`shell` — flow through the same gate when mounting, not just when
installing. Network-only kinds (`http`, `openapi`, `homeassistant`),
hardware-direct kinds (`ble`, `serial`, `adb`), and `mcp http` (no local
process) skip the gate.
"""
from __future__ import annotations

import fnmatch
from dataclasses import dataclass


RISKY_KINDS = {"mcp", "process", "shell"}

# Defaults — what counts as trusted out of the box. Conservative on
# purpose: official registry + the reference servers from the MCP team.
DEFAULT_TRUSTED = (
    "official",
    "npm:@modelcontextprotocol/*",
)


class PolicyDenied(Exception):
    """Raised when policy=='off' refuses an install or risky mount."""


@dataclass
class PolicyDecision:
    allow:  bool
    prompt: bool
    reason: str = ""


class InstallPolicy:
    """Centralized trust gate. One method: `decide(...)`.

    Patterns in `trusted_sources` match either:
      - a bare source name (e.g. `"official"`)  →  matches candidate.source
      - `"<source>:<glob>"` (e.g. `"npm:@modelcontextprotocol/*"`)
        → matches `candidate.source` AND fnmatch on `candidate.id`
    """

    def __init__(self, mode: str = "auto", trusted_sources=None):
        mode = (mode or "auto").lower()
        if mode not in ("auto", "prompt", "off"):
            mode = "auto"
        self.mode = mode
        self.trusted_sources = list(trusted_sources or DEFAULT_TRUSTED)

    def is_trusted(self, source: str, candidate_id: str) -> bool:
        for pat in self.trusted_sources:
            if ":" in pat:
                src, _, glob = pat.partition(":")
                if src == source and fnmatch.fnmatch(candidate_id, glob):
                    return True
            else:
                if pat == source:
                    return True
        return False

    def decide_install(self, source: str, candidate_id: str) -> PolicyDecision:
        if self.mode == "off":
            return PolicyDecision(allow=False, prompt=False,
                                  reason="install_policy=off")
        if self.mode == "prompt":
            return PolicyDecision(allow=True, prompt=True,
                                  reason="install_policy=prompt")
        # mode == "auto"
        if self.is_trusted(source, candidate_id):
            return PolicyDecision(allow=True, prompt=False,
                                  reason="trusted source")
        return PolicyDecision(allow=True, prompt=True,
                              reason="untrusted source — confirming")

    def decide_mount(self, kind: str, source: str = "user", id_: str = "") -> PolicyDecision:
        """Used when DevicesPlugin mounts a bridge directly (not via the
        MCP installer). Only risky kinds gate; everything else is allowed
        silently."""
        if kind not in RISKY_KINDS:
            return PolicyDecision(allow=True, prompt=False, reason="non-risky kind")
        if self.mode == "off":
            return PolicyDecision(allow=False, prompt=False,
                                  reason="install_policy=off")
        if self.mode == "prompt":
            return PolicyDecision(allow=True, prompt=True,
                                  reason="install_policy=prompt")
        if self.is_trusted(source, id_):
            return PolicyDecision(allow=True, prompt=False, reason="trusted")
        return PolicyDecision(allow=True, prompt=True,
                              reason="risky kind — confirming once")

File: cli-ink/src/bridges/policy.ts
/**
 * Trust policy gate. Trusted MCP sources install silently; unknown sources
 * prompt once. `auto` is the default — autonomy without recklessness.
 */

const RISKY_KINDS = new Set(["mcp", "process", "shell"]);

const DEFAULT_TRUSTED = ["official", "npm:@modelcontextprotocol/*"];

export class PolicyDenied extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PolicyDenied";
  }
}

export interface PolicyDecision {
  allow: boolean;
  prompt: boolean;
  reason: string;
}

export type PolicyMode = "auto" | "prompt" | "off";

function fnmatch(pattern: string, value: string): boolean {
  // Convert glob to a regex anchored at both ends.
  const re = new RegExp(
    "^" +
      pattern
        .replace(/[.+^${}()|[\]\\]/g, "\\$&")
        .replace(/\*/g, ".*")
        .replace(/\?/g, ".") +
      "$",
  );
  return re.test(value);
}

export class InstallPolicy {
  mode: PolicyMode;
  trustedSources: string[];

  constructor(mode: string = "auto", trustedSources?: string[]) {
    const m = (mode || "auto").toLowerCase();
    this.mode = (m === "auto" || m === "prompt" || m === "off" ? m : "auto") as PolicyMode;
    this.trustedSources = trustedSources ?? [...DEFAULT_TRUSTED];
  }

  isTrusted(source: string, candidateId: string): boolean {
    for (const pat of this.trustedSources) {
      if (pat.includes(":")) {
        const idx = pat.indexOf(":");
        const src = pat.slice(0, idx);
        const glob = pat.slice(idx + 1);
        if (src === source && fnmatch(glob, candidateId)) return true;
      } else if (pat === source) {
        return true;
      }
    }
    return false;
  }

  decideInstall(source: string, candidateId: string): PolicyDecision {
    if (this.mode === "off") return { allow: false, prompt: false, reason: "install_policy=off" };
    if (this.mode === "prompt") return { allow: true, prompt: true, reason: "install_policy=prompt" };
    if (this.isTrusted(source, candidateId)) return { allow: true, prompt: false, reason: "trusted source" };
    return { allow: true, prompt: true, reason: "untrusted source — confirming" };
  }

  decideMount(kind: string, source = "user", id = ""): PolicyDecision {
    if (!RISKY_KINDS.has(kind)) return { allow: true, prompt: false, reason: "non-risky kind" };
    if (this.mode === "off") return { allow: false, prompt: false, reason: "install_policy=off" };
    if (this.mode === "prompt") return { allow: true, prompt: true, reason: "install_policy=prompt" };
    if (this.isTrusted(source, id)) return { allow: true, prompt: false, reason: "trusted" };
    return { allow: true, prompt: true, reason: "risky kind — confirming once" };
  }
}
